### PR TITLE
feat!: prove transaction inclusion via `ShareProof`

### PR DESCRIPTION
**Pull request by rootulp**
Part of https://github.com/celestiaorg/celestia-core/issues/920

Closes https://github.com/celestiaorg/celestia-core/issues/948, https://github.com/celestiaorg/celestia-core/issues/949

Marked as breaking because protobuf and exported public types have been modified.

## Description

1. Reverts celestia-core specific changes to `TxProof` back to vanilla Tendermint
2. Replace `TxProof` with `ShareProof` in the RPC between celestia-core and celestia-app
3. Extracts `ShareProof` and `RowProof` to new files
4. Improve godocs
5. Improve error handling in validating and verifying proofs

### CODE REVIEW

**Comment by rootulp** [comment]
This PR is almost ready for review. Currently integ test is failing b/c the client used in the test isn't returning something that can be unmarshaled as a share proof so I likely need to update a mock
```
--- FAIL: TestTx (0.03s)
```

File: proto/tendermint/types/types.proto [hunk 1]
```diff
@@ -167,9 +167,9 @@ message BlockMeta {
 
 // TxProof represents a Merkle proof of the presence of a transaction in the Merkle tree.
 message TxProof {
```
Comment (cmwaters) [inline proto/tendermint/types/types.proto:169]: We're not using this at all right?

**Review by cmwaters — APPROVED** [review]
The verification looks good to me although It'd pay to have a few more eyes.

File: light/rpc/client.go [hunk 2]
```diff
@@ -535,18 +535,14 @@ func (c *Client) Tx(ctx context.Context, hash []byte, prove bool) (*ctypes.Resul
...
+	l, err := c.updateLightClientIfNeededTo(ctx, &res.Height)
 	if err != nil {
 		return nil, err
 	}
-
-	valid := res.Proof.VerifyProof()
-	if !valid {
-		err = errors.New("proof for transaction inclusion could not be verified")
-	}
+	fmt.Printf("data hash %v", l.DataHash)
```
[the remark this answers is not in the document]
Comment (rootulp) [reply light/rpc/client.go]: oof yes good catch. Removed in d6ee36a86b133ba12926033b6d690e25077ea644

File: proto/tendermint/types/types.proto (hunk 1 above)
Comment (rootulp) [reply proto/tendermint/types/types.proto:169]: correct

**Comment by rootulp** [comment]
> Do you plan on writing unit tests for the share proof and row proof?

I should but I think those unit tests will need to live in celestia-app. It's a little weird for the types to be defined in this repo and unit tests defined in a separate repo but I don't see how we can generate data squares / transactions / tx inclusion proofs in celestia-core. Will create a follow-up issue in celestia-app.

**Review by SweeXordious — DISMISSED** [review]
LGTM :+1: :+1: 

[optional nit]
I think it would be clearer to add a comment somewhere next to `TxProof` explaining why we're not using it, so that newcomers don't get lost.

**Review by SweeXordious — APPROVED** [review]
:+1: :+1: :+1:

**Review by evan-forbes — APPROVED** [review]
nice, makes sense

**Comment by rootulp** [comment]
Agreed. I attempted to address in https://github.com/celestiaorg/celestia-core/pull/950#issuecomment-1408859767 but may not have been clear

**Pull request merged**